src/txrequest.h

```diff
@@ -0,0 +1,347 @@
+// Copyright (c) 2020 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_TXREQUEST_H
+#define BITCOIN_TXREQUEST_H
+
+#include <crypto/siphash.h>
+#include <uint256.h>
+
+#include <boost/multi_index_container.hpp>
+#include <boost/multi_index/mem_fun.hpp>
+#include <boost/multi_index/ordered_index.hpp>
+
+#include <chrono>
+#include <unordered_map>
+#include <vector>
+
+#include <stdint.h>
+
+
+/** A functor with embedded salt that computes priority of a txid/peer combination.
+ *
+ * The priority is used to order candidates for a given txid; lower priority values are selected first. It has the
+ * following properties:
+ * - Entries are called delayed if they have a reqtime different from std::chrono::microseconds::min(). This is the
+ *   case for entries from inbound non-whitelisted nodes. The priority for delayed entries is always higher than
+ *   non-delayed ones, so non-delayed ones are always preferred.
+ * - Within the set of delayed peers and within the set of non-delayed peers:
+ *   - The very first announcement for a given txid is always lowest.
+ *   - Announcements after the first one are ordered uniformly randomly (using a salted hash of txid/peer).
+ */
+class PriorityComputer {
+    const uint64_t m_k0, m_k1;
+public:
+    explicit PriorityComputer(bool deterministic);
+    uint64_t operator()(const uint256& txid, uint64_t peer, bool delayed, bool first) const
+    {
+        uint64_t low_bits = 0;
+        if (!first) {
+            low_bits = CSipHasher(m_k0, m_k1).Write(txid.begin(), txid.size()).Write(peer).Finalize() >> 1;
+        }
+        return low_bits | uint64_t{delayed} << 63;
+    }
+};
+
+/** Data structure to keep track of, and schedule, transaction downloads from peers.
+ *
+ * === High level behavior ===
+ *
+ * Transaction downloads are attempted in random order, first choosing among non-delayed peers, then delayed
```

Comment:
See #19988. The "delayed"/"nondelayed" distinction doesn't exist anymore. There is an explicit separate boolean to `ReceivedInv` now for preferred/nonpreferred, and a way to specify reqtime explicitly. (in particular, because txid peers get a delay, even though they're not preferred like outbounds are - so these concepts needed to be split).